Python/executor.c

```diff
@@ -0,0 +1,125 @@
+
+#include "pycore_call.h"
+#include "pycore_ceval.h"
+#include "pycore_dict.h"
+#include "pycore_emscripten_signal.h"
+#include "pycore_intrinsics.h"
+#include "pycore_long.h"
+#include "pycore_object.h"
+#include "pycore_opcode_metadata.h"
+#include "pycore_opcode_utils.h"
+#include "pycore_pyerrors.h"
+#include "pycore_range.h"
+#include "pycore_sliceobject.h"
+#include "pycore_uops.h"
+
+#include "ceval_macros.h"
+
+
+#undef ASSERT_KWNAMES_IS_NULL
+#define ASSERT_KWNAMES_IS_NULL() (void)0
+
+#undef DEOPT_IF
+#define DEOPT_IF(COND, INSTNAME) \
+    if ((COND)) {                \
+        goto deoptimize;         \
+    }
+
+#undef ENABLE_SPECIALIZATION
+#define ENABLE_SPECIALIZATION 0
+
+
+_PyInterpreterFrame *
+_PyUopExecute(_PyExecutorObject *executor, _PyInterpreterFrame *frame, PyObject **stack_pointer)
+{
+#ifdef Py_DEBUG
+    char *uop_debug = Py_GETENV("PYTHONUOPSDEBUG");
+    int lltrace = 0;
+    if (uop_debug != NULL && *uop_debug >= '0') {
+        lltrace = *uop_debug - '0';  // TODO: Parse an int and all that
+    }
+    #define DPRINTF(level, ...) \
+        if (lltrace >= (level)) { fprintf(stderr, __VA_ARGS__); }
+#else
+    #define DPRINTF(level, ...)
+#endif
+
+    DPRINTF(3,
+            "Entering _PyUopExecute for %s (%s:%d) at byte offset %ld\n",
+            PyUnicode_AsUTF8(_PyFrame_GetCode(frame)->co_qualname),
+            PyUnicode_AsUTF8(_PyFrame_GetCode(frame)->co_filename),
+            _PyFrame_GetCode(frame)->co_firstlineno,
+            2 * (long)(frame->prev_instr + 1 -
+                   (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive));
+
+    PyThreadState *tstate = _PyThreadState_GET();
+    _PyUOpExecutorObject *self = (_PyUOpExecutorObject *)executor;
+
+    CHECK_EVAL_BREAKER();
+
+    OBJECT_STAT_INC(optimization_traces_executed);
+    _Py_CODEUNIT *ip_offset = (_Py_CODEUNIT *)_PyFrame_GetCode(frame)->co_code_adaptive;
+    int pc = 0;
+    int opcode;
+    int oparg;
+    uint64_t operand;
+
+    for (;;) {
+        opcode = self->trace[pc].opcode;
+        oparg = self->trace[pc].oparg;
+        operand = self->trace[pc].operand;
+        DPRINTF(3,
+                "%4d: uop %s, oparg %d, operand %" PRIu64 ", stack_level %d\n",
+                pc,
+                opcode < 256 ? _PyOpcode_OpName[opcode] : _PyOpcode_uop_name[opcode],
+                oparg,
+                operand,
+                (int)(stack_pointer - _PyFrame_Stackbase(frame)));
+        pc++;
+        OBJECT_STAT_INC(optimization_uops_executed);
+        switch (opcode) {
+
+#include "executor_cases.c.h"
+
+            default:
+            {
+                fprintf(stderr, "Unknown uop %d, operand %" PRIu64 "\n", opcode, operand);
+                Py_FatalError("Unknown uop");
+            }
+
+        }
+    }
+
+unbound_local_error:
+    _PyEval_FormatExcCheckArg(tstate, PyExc_UnboundLocalError,
+        UNBOUNDLOCAL_ERROR_MSG,
+        PyTuple_GetItem(_PyFrame_GetCode(frame)->co_localsplusnames, oparg)
+    );
```

Comment:
I think it would be fine to call `_PyEval_FormatExcUnbound` here and at the corresponding point in ceval.c, then you won't have to move `UNBOUNDLOCAL_ERROR_MSG` into the header file (which feels a little unprincipled).

Reply:
I can do that, sure.

FWIW though, `NAME_ERROR_MSG` is already in `ceval_macros.h`, so it's not *super* out of place there. In fact, it might be nice to put `NAME_ERROR_MSG`, `UNBOUNDFREE_ERROR_MSG`, `UNBOUNDLOCAL_ERROR_MSG` together in the same place.

Reply:
Yeah, all together in the same place is good, so then I vote for moving all to the header and not changing the code at the label.